src/System.Resources.Binary.Reader/src/System/Resources/Binary/BinaryResourceReader.cs

```diff
@@ -0,0 +1,133 @@
+
+                        value = bf.Deserialize(reader.BaseStream);
+
+                        long bytesRead = reader.BaseStream.Position - originalPosition;
+
+                        // Ensure BF read what we expected.
+                        if (bytesRead != length)
+                        {
+                            throw new BadImageFormatException(SR.Format(SR.BadImageFormat_ResourceDataLengthInvalid, length));
+                        }
+                        break;
+                    }
+                case SerializationFormat.TypeConverterByteArray:
+                    {
+                        // read length
+                        int length = reader.Read7BitEncodedInt();
+                        if (length < 0)
+                        {
+                            throw new BadImageFormatException(SR.Format(SR.BadImageFormat_ResourceDataLengthInvalid, length));
+                        }
+
+                        byte[] data = reader.ReadBytes(length);
+
+                        TypeConverter converter = TypeDescriptor.GetConverter(type);
+
+                        if (converter == null)
+                        {
+                            throw new TypeLoadException(SR.Format(SR.TypeLoadException_CannotLoadConverter, type));
+                        }
+
+                        value = converter.ConvertFrom(data);
+                        break;
+                    }
+                case SerializationFormat.TypeConverterString:
+                    {
+                        string stringData = reader.ReadString();
+
+                        TypeConverter converter = TypeDescriptor.GetConverter(type);
+
+                        if (converter == null)
+                        {
+                            throw new TypeLoadException(SR.Format(SR.TypeLoadException_CannotLoadConverter, type));
+                        }
+
+                        value = converter.ConvertFrom(stringData);
+                        break;
+                    }
+                case SerializationFormat.Stream:
+                    {
+                        // read length
+                        int length = reader.Read7BitEncodedInt();
+                        if (length < 0)
+                        {
+                            throw new BadImageFormatException(SR.Format(SR.BadImageFormat_ResourceDataLengthInvalid, length));
+                        }
+                        Stream stream;
+
+                        if (reader.BaseStream is UnmanagedMemoryStream ums)
+                        {
+                            // For the case that we've memory mapped in the .resources
+                            // file, just return a Stream pointing to that block of memory.
+                            unsafe
+                            {
+                                stream = new UnmanagedMemoryStream(ums.PositionPointer, length, length, FileAccess.Read);
```

Comment:
Looks like a bug in the original `ResourceReader` code. The `ResourceReader` ctor assumes that any instance of an `UnmanagedMemoryStream` passed in to its ctor must correspond to memory that will never be freed for the lifetime of the application, because a non-lifetime-tracked pointer to that memory is being exposed to the outside world via this method. While this is true for resources loaded from assemblies (where the assembly isn't in an unloadable context), this is not generally a safe assumption to make.

I don't think we need to address this now since this code has existed for 20 years without issue, but it's definitely a bug.

Reply:
Agreed this is a bad assumption.  UMW doesn't provide any way to rent out a ref count to its pointer.  Probably need a better check to make sure the UMS points to the embedded resources if we wanted to avoid this.

Reply:
> I don't think we need to address this now since this code has existed for 20 years without issue, but it's definitely a bug.

This was fixed recently: https://github.com/dotnet/coreclr/pull/22925/

Reply:
I see, we can't take advantage of that in the netstandard build, but can in the corelib version.